Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp

```diff
@@ -219,46 +369,53 @@ void DiscusMultipleScatteringCorrection::exec() {
   const MatrixWorkspace_sptr inputWS = getProperty("InputWorkspace");
-  MatrixWorkspace_sptr SQWS = getProperty("SofqWorkspace");
-  // avoid repeated conversion of bin edges to points inside loop by converting to point data
-  if (SQWS->isHistogramData()) {
-    // flat interpolation is later used on S(Q) so convert to points by assigning Y value to LH bin edge
-    MatrixWorkspace_sptr SQWSPoints = API::WorkspaceFactory::Instance().create(SQWS, SQWS->getNumberHistograms(),
-                                                                               SQWS->blocksize(), SQWS->blocksize());
-    SQWSPoints->setSharedY(0, SQWS->sharedY(0));
-    SQWSPoints->setSharedE(0, SQWS->sharedE(0));
-    std::vector<double> newX = SQWS->histogram(0).dataX();
-    newX.pop_back();
-    SQWSPoints->setSharedX(0, HistogramData::Points(newX).cowData());
-    SQWS = SQWSPoints;
-  }
+
+  m_SQWS = getProperty("StructureFactorWorkspace");
+  prepareStructureFactor();
 
   MatrixWorkspace_sptr sigmaSSWS = getProperty("ScatteringCrossSection");
   if (sigmaSSWS)
-    m_sigmaSS = std::make_shared<HistogramData::Histogram>(sigmaSSWS->histogram(0));
-
-  double lambdamin, lambdamax;
-  getXMinMax(*inputWS, lambdamin, lambdamax);
-  m_SQHist = std::make_shared<HistogramData::Histogram>(SQWS->histogram(0));
-  auto QSQHist = prepareQSQ(2 * M_PI / lambdamin);
-
-  const auto inputNbins = static_cast<int>(inputWS->blocksize());
-  int nlambda = getProperty("NumberOfWavelengthPoints");
-  if (isEmpty(nlambda) || nlambda > inputNbins) {
-    if (!isEmpty(nlambda)) {
-      g_log.warning() << "The requested number of wavelength points is larger "
-                         "than the spectra size. "
-                         "Defaulting to spectra size.\n";
+    m_sigmaSS = std::make_shared<DataObjects::Histogram1D>(sigmaSSWS->getSpectrum(0));
+
+  // for inelastic we could calculate the qmax based on the min\max w in the S(Q,w) but that
+  // would bake as assumption that S(Q,w)=0 beyond the limits of the supplied data
+  double qmax = std::numeric_limits<float>::max();
+  EFixedProvider efixed(*inputWS);
+  m_EMode = efixed.emode();
+  g_log.information("EMode=" + DeltaEMode::asString(m_EMode) + " detected");
+  if (m_EMode == Kernel::DeltaEMode::Elastic) {
+    double kmin, kmax;
+    getXMinMax(*inputWS, kmin, kmax);
+    qmax = 2 * kmax;
+  }
+  m_QSQWS = prepareQSQ(qmax);
+  size_t totalPoints = 0;
+  for (size_t i = 0; i < m_QSQWS->getNumberHistograms(); i++) {
+    totalPoints += m_QSQWS->histogram(i).size();
+  }
+
+  m_simulateEnergiesIndependently = getProperty("SimulateEnergiesIndependently");
+  // call this function with dummy efixed to determine total possible simulation points
+  const auto inputNbins = generateInputKOutputWList(-1.0, inputWS->points(0).rawData()).size();
+
+  int nSimulationPointsInt = getProperty("NumberOfSimulationPoints");
+  size_t nSimulationPoints = static_cast<size_t>(nSimulationPointsInt);
```

Comment:
This can be done as a single step:
```c++
auto nSimulationPoints = static_cast<size_t>(getProperty("NumberOfSimulationPoints"));
```

Reply:
I did try this initially but it gives this error when you supply a value for this parameter:
`Attempt to assign property NumberOfSimulationPoints to incorrect type. Expected type uint64_t`
It seems like the result of a call to getProperty needs to be assigned to an int variable initially because the parameter is set up as an int in the algorithm's Init method.
Having a size_t parameter seems quite hard work in Mantid and is something I've come across before on other algorithms. If you instead set it up as a size_t in init, it makes supplying values a bit of a pain because there aren't size_t literals. So hard to write a C++ unit test for example if the parameter is set up as size_t